### ISSUE TITLE
Pull based on partial array match

### ISSUE THREAD

**Issue by ndarilek** [jagi/meteor-astronomy#275]
I have the following specification:

```
User.extend

  fields:
    pushNotificationTokens:
      type: "array"
      nested:
        name: "PushNotificationToken"
        fields:
          token:
            type: "string"
            default: -> uuid()
          createdAt:
            type: "date"
            default: -> new Date()
      default: -> []
```

If someone submits a request with a token, I'd like to remove it from the list of valid tokens so it can't be used again.

Unfortunately, `user.pull("pushNotificationTokens", {token: "value of valid token"})` doesn't appear to remove the token. Initially I thought the array had to be an exact match, but looking at the Mongo $pull docs which I assume this is modeled off of, I now don't get that impression.

Is there anything obviously wrong with my query, and is there some way I can get a look at the underlying Mongo query to figure out what might be going wrong?

Thanks.

**Comment by jagi** [comment]
Value (element) used in the `pull` method has to be exact match of the array element that you want to remove. So it not only have to contain `token` but also the `createdAt` field. The best way to achieve that is first look for the element with a given token and later remove it.

```js
var element = _.find(user.pushNotificationTokens, function(element) {
  return element.token === token;
});
user.pull('pushNotificationTokens', element);
```

**Comment by willrbc** [comment]
Interesting,  I'm having a similar problem,  although I have the exact match of a nested array available.

I've tried your above solution and it's no different.

In the example below I know that the contact I am trying to delete is at index 10 of the array so I am doing a comparison which returns true whether I use

```javascript
    var company = Template.parentData(2).company;
    var c = _.find(company.contacts, function (item) {
      return item.contactId === tmpl.data.contact.contactId;
    });
    console.log(company.contacts[10], c, company.contacts[10] === c);
    company.pull("contacts", c);
    var company = Template.parentData(2)
      .company;
    var c = _.find(company.contacts, function (item) {
      return item.email === tmpl.data.contact.email; //email is unique in this project
    });
    console.log(company.contacts[10] === c); //returns 'true' so there is exact match
    console.log(company.contacts[10] === tmpl.data.contact); //also returns 'true' so there is exact match
    
    company.pull("contacts", c);
    console.log(company.contacts[10], company.contacts); //contact 'c' is still in the array  
 company.pull("contacts", tmpl.data.contact);
    console.log(company.contacts[10], company.contacts); //contact from template data is still in the array  
```

**Comment by jagi** [comment]
@willrbc can you create reproduction repository? I would like to have ability to investigate it.

**Comment by willrbc** [comment]
I hope this repo makes sense.  You will notice the issue when you try to add and then immediately remove a Phone number from a contact.

The weird this is, it works after you save and reopen the contact and then remove a number, it just doesn't work when the Class is open in the ReactiveVar

**Comment by jagi** [comment]
The problem with your code is that you're pushing empty phone number first and later trying to remove it. And it's not possible in Astronomy 1.x to do that. Performing two operations on the same field is not possible. Unfortunately I can't do anything about it. And that's why in Astronomy 2.0 I've change entirely how modification system works and this problems does not occur any more. What I would suggest for now is instead pushing just setting entire `telephone` array. You can perform multiple set operations on the same field but not `push` and `pull` operations on the same field.

**Comment by willrbc** [comment]
Just so I understand you correctly, when you say performing two operations is not possible, do you mean two operations between saves? If I was saving the document after each push and before the pull it would work?  I've already used the set workaround, but thought I might clarify this issue. 

Thanks for your help.

**Comment by jagi** [comment]
@willrbc To be precise you can't use two different modifiers on the same field between saves.
```js
doc.set('phones', []);
doc.push('phones', {number: 123});
doc.save(); // It will not work
```
However you can do:
```js
doc.set('phones', []);
doc.set('phones', [{number: 123}]);
doc.save();
```
It limitation of MongoDB, that for which I found the solution in Astronomy 2.0. It just doesn't use Mongo modifiers to register changes. You can freely change a document, and later when saving Astronomy will try to generate the Mongo modifier that is as minimal as possible.

**Comment by jagi** [comment]
Closing this issue as it's limitation of 1.x and is solved in 2.0
